research/job_referencing/index.md

```diff
@@ -0,0 +1,217 @@
+
+client = await Client.connect("localhost:7233")
+
+# Run a worker for the workflow
+async with Worker(
+  client,
+  task_queue="project-sha-event-queue",
+  workflows=[COPRBuildHandler],
+  activities=[submit_build_srpm,
+              submit_build_copr,
+              update_ux]
+  ):
+
+  # While the worker is running, use the client to start the workflow.
+  # Note, in many production setups, the client would be in a completely
+  # separate process from the worker.
+  result = await client.execute_workflow(
+    COPRBuildHandler.run,
+    id="copr-build-for-project-and-sha",
+    task_queue="project-sha-event-copr-build-queue",
+    execution_timeout=timedelta(minutes=60)
+```
+
+```python
+from temporalio import workflow
+from temporalio.client import Client
+from temporalio.worker import Worker
+
+@activity.defn
+async def submit_build_srpm():
+  ...
+
+@activity.defn
+async def submit_build_copr():
+  ...
+
+@activity.defn
+async def update_ux():
+  ...
+
+class COPRBuildHandler:
+
+  def __init__(self) -> None:
+    self._srpm_builded = False
+    self._copr_builded = False
+
+  @reacts_to(event=CoprBuildEndEvent)
+  @workflow.signal
+  def submit_srpm_build_ended():
+    if event.chroot == COPR_SRPM_CHROOT:
+      self._srpm_builded = True
+
+  @reacts_to(event=CoprBuildEndEvent)
+  @workflow.signal
+  def submit_copr_build_ended():
+    if event.chroot != COPR_SRPM_CHROOT:
+      self._copr_builded = True
+
+  @workflow.run
+  async def run(self):
+
+    srpm_activity_handle = workflow.start_activity(
+      submit_build_srpm,
+      start_to_close_timeout=timedelta(seconds=1),
+      ...
+      )
+    await workflow.wait_condition(lambda: self._srpm_builded)
+
+    update_ux_handle = workflow.start_activity(
+      update_ux,
+      start_to_close_timeout=timedelta(seconds=1),
+      ...
+      )
+
+    build_copr_handle = workflow.start_activity(
+      build_copr,
+      start_to_close_timeout=timedelta(seconds=1),
+      ...
+      )
+    await workflow.wait_condition(lambda: self._copr_builded)
+
+    update_ux_handle = workflow.start_activity(
+      update_ux,
+      start_to_close_timeout=timedelta(seconds=1),
+      ...
+      )
+```
+
+As far as I understand from the `temporalio` documentation, the framework is in charge of putting this `CoprBuildHandler workflow` in a sleeping queue while it is waiting for a signal and waking up it later. So we don't need to exit the workflow and the order of the activities is well visible in the `workflow` code.
+
+There is another small improvement I see using a solution like this one instead of doing what we are already doing.
+I don't think that the `CoprBuildEndHandler` should deal with the `TestingFarmHandler` and viceversa, this is dangerous, and the more dependencies we will introduce the more we will need this dangerous cross referencing.
+
+Something above them should be able to orchestrate and know the jobs, and I think a `workflow` is exactly what we are missing here.
```

Comment:
We can maybe discuss this before spending more time on that, but do you have an idea how we should transition to this (and how easy/hard this might be)?

Reply:
Thinking about that more, what about trying to use this e.g. for Copr builds (including all the steps) and run it inside the current celery task?

Or try this with something else to see how this can work in our use cases.

Reply:
I am starting to like this framework. This is actually quite close to what we need: https://learn.temporal.io/tutorials/python/subscriptions/

Reply:
> Thinking about that more, what about trying to use this e.g. for Copr builds (including all the steps) and run it inside the current celery task?
> 
> Or try this with something else to see how this can work in our use cases.

I will think more about it. I don't think this can be done just using `asyncio`; we will end up, I think, having a lot of long-running workers and I think this is not ideal. But maybe I am wrong, I don't really know how Celery copes with asyncio.

Reply:
You are right, this combination might not be doable.